core/services/ocr2/plugins/s4/plugin.go

```diff
@@ -83,48 +89,81 @@ func (c *plugin) Observation(ctx context.Context, _ types.ReportTimestamp, query
 		return nil, errors.Wrap(err, "failed to DeleteExpired in Observation()")
 	}
 
-	queryRows := make([]*s4.Row, 0)
+	returnObservation := func(rows []*s4.Row) (types.Observation, error) {
+		promReportingPluginsObservationRowsCount.WithLabelValues(c.config.ProductName).Set(float64(len(rows)))
+		return MarshalRows(convertRows(rows))
+	}
+
 	unconfirmedRows, err := c.orm.GetUnconfirmedRows(c.config.MaxObservationEntries, pg.WithParentCtx(ctx))
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to GetUnconfirmedRows in Observation()")
 	}
 
-	if uint(len(unconfirmedRows)) < c.config.MaxObservationEntries {
-		versionRows, err := UnmarshalQuery(query)
+	if uint(len(unconfirmedRows)) >= c.config.MaxObservationEntries {
+		return returnObservation(unconfirmedRows[:c.config.MaxObservationEntries])
+	}
+
+	maxRemainingRows := int(c.config.MaxObservationEntries) - len(unconfirmedRows)
+	remainingRows := make([]*s4.Row, 0)
+
+	queryRows, err := UnmarshalQuery(query)
+	if err != nil {
+		c.logger.Errorw("Failed to unmarshal query (likely malformed)", "err", err)
+	} else {
+		snapshot, err := c.orm.GetSnapshot(s4.NewFullAddressRange(), pg.WithParentCtx(ctx))
 		if err != nil {
-			c.logger.Errorw("Failed to UnmarshalQuery, likely data is malformed", "err", err)
+			c.logger.Errorw("ORM GetSnapshot error", "err", err)
 		} else {
-			maxObservationRows := int(c.config.MaxObservationEntries) - len(unconfirmedRows)
-			for _, vr := range versionRows {
-				address := UnmarshalAddress(vr.Address)
-				row, err := c.orm.Get(address, uint(vr.Slotid), pg.WithParentCtx(ctx))
-				if err == nil && row.Version > vr.Version {
-					queryRows = append(queryRows, row)
-				} else if err != nil && !errors.Is(err, s4.ErrNotFound) {
-					c.logger.Errorw("ORM Get error", "err", err)
+			type rkey struct {
+				address *utils.Big
+				slotID  uint
+			}
+
+			snapshotVersionsMap := snapshotToVersionMap(snapshot)
+			toBeAdded := make([]rkey, 0)
+			for _, qr := range queryRows {
+				address := UnmarshalAddress(qr.Address)
+				k := key{address: address.String(), slotID: uint(qr.Slotid)}
+				if version, ok := snapshotVersionsMap[k]; ok && version >= qr.Version {
+					toBeAdded = append(toBeAdded, rkey{address: address, slotID: uint(qr.Slotid)})
+					delete(snapshotVersionsMap, k)
 				}
-				if len(queryRows) >= maxObservationRows {
-					break
+			}
+
+			if len(toBeAdded) > maxRemainingRows {
```

Comment:
Maybe add a snapshot length check? If query has the same number of entries as your snapshot you can short-circuit and exit here because you won't find anything else to be added. That would be the most common scenario.

Reply:
Not true, because query and snapshot may have different non-intersecting records.

Reply:
You're right, my bad!